File: LeetCodeProblems/HashingOrArrays/ContainsDuplicateIIHashSetSolution_219.cs
using LeetCodeProblems.Interfaces.Easy;

namespace LeetCodeProblems.HashingOrArrays
{
    public class ContainsDuplicateIIHashSetSolution_219 : IContainsDuplicateII_219
    {
        /// <summary>
        /// Given an integer array nums and an integer k, return true if there are two distinct indices i and j in the array such that nums[i] == nums[j] and abs(i - j) <= k.
        /// </summary>
        /// <param name="nums"></param>
        /// <param name="k"></param>
        /// <returns></returns>
        public bool ContainsNearbyDuplicate(int[] nums, int k)
        {
            if (nums.Length == 0 || k == 0) return false;

            var hashSet = new HashSet<int>();

            for ( int i = 0; i < nums.Length; i++)
            {
                if (hashSet.Contains(nums[i])) return true;
                else hashSet.Add(nums[i]);
                if (i >= k) hashSet.Remove(nums[i-k]);
            }
            return false;
        }
    }
}

File: LeetCodeProblems.VisualBasic/HashingOrArrays/ContainsDuplicateIIHashSetSolution_219.vb
Imports LeetCodeProblems.Interfaces.Easy

Namespace HashingOrArrays
    Public Class ContainsDuplicateIIHashSetSolution_219
        Implements IContainsDuplicateII_219


        ''' <summary>
        ''' this uses a hashmap like the C# implementation, 
        ''' but instead of treating i as the right boundary, 
        ''' it assumes i is the left boundary and i+k is the right boundary
        ''' </summary>
        ''' <returns></returns>
        Public Function ContainsNearbyDuplicate(nums() As Integer, k As Integer) As Boolean Implements IContainsDuplicateII_219.ContainsNearbyDuplicate
            If (nums.Length = 0 OrElse k = 0) Then Return False

            Dim hashMap As New HashSet(Of Integer)()

            For i = 0 To Math.Min(k - 1, nums.Length - 1)
                If hashMap.Contains(nums(i)) Then Return True

                hashMap.Add(nums(i))
            Next

            For i = 0 To nums.Length - 1 - k

                Dim targetOffset = i + k
                If hashMap.Contains(nums(targetOffset)) Then Return True
                hashMap.Remove(nums(i))
                hashMap.Add(nums(targetOffset))
            Next

            Return False
        End Function
    End Class

End Namespace

